lib/schema/duplicate.js

```diff
@@ -0,0 +1,48 @@
+'use strict';
+
+var amoeba = require('amoeba');
+var except = amoeba.except;
+var misc = require('../misc.js');
+
+var idHashFieldMap = {};
+
+exports.registerFieldsForDuplicator = function (type, idHashFields) {
+  let all = ['_userId', 'deviceId', 'time', 'type'];
+  if (idHashFields && idHashFields.length > 0) {
+    all = all.push(...idHashFields);
+  }
+  if (idHashFieldMap[type] == null) {
+    idHashFieldMap[type] = all;
+  } else {
+    throw except.IAE(
+      'Id hash fields for type[%s] already defined[%j], cannot set[%j]',
+      type,
+      idHashFieldMap[type],
+      idHashFields
+    );
+  }
+};
+
+exports.generateHash = function (datum) {
+  if (typeof datum === 'string') {
+    return datum;
+  }
+
+  var idHashFields = idHashFieldMap[datum.type];
+  if (!idHashFields) {
+    return '';
+  }
+
+  var vals = new Array(idHashFields.length);
+  for (var i = 0; i < idHashFields.length; ++i) {
+    vals[i] = datum[idHashFields[i]];
```

Comment:
I believe all time fields are expected to be formatted as `RFC3339Nano` in platform. In this case you need to make sure that all time fields are actually `Date` ( I believe they already are if `filterDatumForMongo` is called on the datum before passing it to this and you need to call the date's `toISOString()` method to mimic platform's behaviour for dates instead of relying on the implicit `toString` method call.

Reply:
updates made @lostlevels - thanks for catching that